kernel/src/net/socket/ip/stream/mod.rs

```diff
@@ -570,14 +608,7 @@ impl Socket for StreamSocket {
                 tcp_congestion.set(congestion);
             },
             tcp_maxseg: MaxSegment => {
-                // It will always return the default MSS value defined above for an unconnected socket
-                // and always return the actual current MSS for a connected one.
-
-                // FIXME: how to get the current MSS?
-                let maxseg = match self.state.read().as_ref() {
-                    State::Init(_) | State::Listen(_) | State::Connecting(_) => DEFAULT_MAXSEG,
-                    State::Connected(_) => options.tcp.maxseg(),
-                };
+                let maxseg = options.tcp.maxseg();
```

Comment:
Why changing the maxseg option here? It seems unrelated to the bug?

Reply:
I have checked the [Linux implementation] (https://elixir.bootlin.com/linux/v6.11.2/source/net/ipv4/tcp.c#L4054). Nowhere does it implement the original comments:
```
                // It will always return the default MSS value defined above for an unconnected socket
                // and always return the actual current MSS for a connected one.
```
"It will always return the default MSS value defined above for an unconnected socket" This just looks like a false claim, and it is _not_ the Linux behavior. So I'm confused as to how this code is introduced.

The reason I change this is because I searched all `state.read()` and tried to replace it with `read_updated_state()`. However, I can't replace this with the MAXSEG option because if I replace it, there will be a chance of a deadlock (note that we have already held the write lock of `option` at this point).

Reply:
> So I'm confused as to how this code is introduced.

I remember this code is introduced to pass a gvisor test, but I haven't checked the exact Linux behavior at that time.

But since the gvisor socket test is not added now, the change is acceptable for me since CI is not broken.

Reply:
> But since the gvisor socket test is not added now, the change is acceptable for me since CI is not broken.

Even if the original logic is correct (but again, [_it isn't_](https://elixir.bootlin.com/linux/v6.11.2/source/net/ipv4/tcp.c#L4054)), as a long-term solution we need to reorder the locking order of `state` and `options`. This is because `state` needs to be a mutex, but we might want to keep `options` a spin lock.

However, it's a bigger and irrelevant problem, let's leave this as future work.